quipucords/scanner/normalizer.py

```diff
@@ -0,0 +1,225 @@
+"""normalizer module."""
+
+from dataclasses import dataclass
+from logging import getLogger
+from typing import Any
+
+logger = getLogger(__name__)
+
+
+class NormalizerMeta(type):
+    """
+    Type for Normalizer classes.
+
+    Customize class creation to provide the necessary magic for Normalizers.
+
+    This metaclass is meant to be used ONLY on BaseNormalizer.
+    """
+
+    def __new__(cls, name, bases, attrs):
+        """Blueprint for Normalizer classes."""
+        # we need fields to be accessible as a class attribute. This is the WHOLE reason
+        # for the existence of this metaclass (I wish python had a simpler api for this
+        # type of thing).
+        # Usually just creating a attribute on a base class would be enough
+        # for this behavior, but not when you are dealing with dicts or lists.
+        # (Creating 'fields' as a normal attribute on 'BaseNormalizer' class would
+        # result on it being shared between the base class and its subclasses.)
```

Comment:
I don't understand the part about dicts and lists, but can't we just have a [base object implementing `__getattr__`](https://docs.python.org/3/reference/datamodel.html#object.__getattr__)? That will be called when you try to access `object.field_name`, assuming there's no `field_name` in `object`, so you can check in `object.fields` or whatever before failing.

Reply:
I think I understand the dicts lists comment. This depends on when attribute is defined - it if is defined as class attribute (outside of init or any method), then yes, it will be shared between all instances. If it is defined in method, it is exclusive to instance:

```python
class A:
    class_fields = []
    def __init__(self):
        self.fields = []

    def add_to_fields(self, value):
        self.fields.append(value)
        print("{}.fields of obj {} is set to {}".format(
            self.__class__.__name__, str(id(self)), str(self.fields)
        ))

    def add_to_class_fields(self, value):
        self.class_fields.append(value)
        print("{}.class_fields of obj {} is set to {}".format(
            self.__class__.__name__, str(id(self)), str(self.class_fields)
        ))


class B(A):
    pass

one = B()
two = B()

one.add_to_fields(1)
two.add_to_fields(2)

one.add_to_class_fields(1)
two.add_to_class_fields(2)
```
```
B.fields of obj 139825975519056 is set to [1]
B.fields of obj 139825975519120 is set to [2]
B.class_fields of obj 139825975519056 is set to [1]
B.class_fields of obj 139825975519120 is set to [1, 2]
```

Reply:
Also, FactMapper depends on `fields` existing at class level.

Reply:
I gave it a little thought and there _might_ be three ways to avoid metaclass:

1. Keep `fields` shared class attribute, but instead of it being key -> fact mapping, make it child class "path" (say: `tuple(self.__class__.mro())`) -> key -> fact multi-level mapping.

2. Let `FactMapper.__set_name__` set normalizer class `fields`.

3. Make `FactMapper` lazy - they would not look into normalizer `fields` until specific field is actually requested.

Now, all three of them have major drawbacks, and I am not convinced these drawbacks aren't making them actually _worse_ than using metaclass...

So let's see if anyone else has something to add, but I think we will leave that part as is. At least now I have better grasp _why_ this metaclass exists :+1:

Reply:
2. seems a pretty solid option.

I'm not sure I understood 1, though 😅

Reply:
got it

Reply:
I wrote a small test to confirm the behavior I was expecting is actually happening at class construction (should've written it in the first place).

Test is passing
```python
def test_multiple_normalizers():
    """Ensure one normalizer class does not leak info to another."""
    assert BaseNormalizer.fields == {}

    class N1(BaseNormalizer):
        fact1 = FactMapper("some_raw_fact", str)

    class N2(BaseNormalizer):
        fact2 = FactMapper("another_raw_fact", str)

    assert BaseNormalizer.fields == {}
    assert N1.fields == {"fact1": N1.fact1}
    assert N2.fields == {"fact2": N2.fact2}
```

Will be a good benchmark while I experiment other approaches.

Great thing about getting rid of the metaclass is making BaseNormalizer an abstract class 😬

Reply:
https://github.com/quipucords/quipucords/pull/2492/files#r1348003229